src/GenSequence.ts

```diff
@@ -282,6 +286,33 @@ export function first<T>(fn: (t: T) => boolean, defaultValue: T, i: Iterable<T>)
     return defaultValue;
 }
 
+export function max<T>(i: Iterable<T>): Maybe<T> {
+    let sawValue: boolean = false;
+    const value: Maybe<T> = reduce((p: T, c: T) => {
+        if (!sawValue) {
+            sawValue = true;
+        }
+        
+        if (isNonValue(p) || c > p) {
+            return c;
+        }
+
+        return p;
+    },
+    undefined, 
+    i);
+
+    if (!sawValue) {
+        throw new Error("the sequence is empty.");
+    }
+    
+    if (isNonValue(value)) {
+        return undefined;
+    }
+
+    return value;
+}
+
```

Comment:
The C# linq expression exposes a transform (map basically), but it seems like it just gives  you the equivelent of this:
```javascript
const seq = genSequence([{name="joe", age=15}, {name="nancy", age=43}]);
const m = seq
    .map(v => v.age)
    .max()
```
 That leaves you with m === 43, but you don't know which object that is

What you proposed is better because you get the object back and know all the rest of the information about the object.  Another (slightly cleaner?) way to do what you were proposing might be
`max<T, U>(valueSelector: (T) => U): T`
so you could do this
```javascript
const seq = genSequence([{name="joe", age=15}, {name="nancy", age=43}]);
const m = seq
    .max(v => v.age)
```
so m.age === 43, and m.name === "nancy".

The version you spec'd would look like this:
```javascript
const seq = genSequence([{name="joe", age=15}, {name="nancy", age=43}]);
const m = seq
    .max((c, p) => c.age > p.age)
```
And would have some strange side effects in that you could use the max() funciton to find the min  value if you write the gt function backwards.

Thoughts?